File: src/main/java/ru/citeck/ecos/model/num/api/records/NumTemplateRecordsDao.java
package ru.citeck.ecos.model.num.api.records;

import ecos.com.fasterxml.jackson210.annotation.JsonProperty;
import ecos.com.fasterxml.jackson210.annotation.JsonValue;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.apache.commons.lang.StringUtils;
import org.jetbrains.annotations.Nullable;
import org.springframework.stereotype.Component;
import ru.citeck.ecos.commons.data.ObjectData;
import ru.citeck.ecos.commons.data.entity.EntityWithMeta;
import ru.citeck.ecos.commons.json.Json;
import ru.citeck.ecos.commons.json.YamlUtils;
import ru.citeck.ecos.commons.utils.TmplUtils;
import ru.citeck.ecos.events2.type.RecordEventsService;
import ru.citeck.ecos.model.EcosModelApp;
import ru.citeck.ecos.model.lib.num.dto.NumTemplateDef;
import ru.citeck.ecos.model.num.dto.NumTemplateDto;
import ru.citeck.ecos.model.num.dto.NumTemplateWithMetaDto;
import ru.citeck.ecos.model.num.service.NumTemplateService;
import ru.citeck.ecos.model.utils.LegacyRecordsUtils;
import ru.citeck.ecos.records2.RecordMeta;
import ru.citeck.ecos.records2.RecordRef;
import ru.citeck.ecos.records2.graphql.meta.value.MetaField;
import ru.citeck.ecos.records2.predicate.model.Predicate;
import ru.citeck.ecos.records2.request.delete.RecordsDelResult;
import ru.citeck.ecos.records2.request.delete.RecordsDeletion;
import ru.citeck.ecos.records2.request.mutation.RecordsMutResult;
import ru.citeck.ecos.records2.request.query.RecordsQuery;
import ru.citeck.ecos.records2.request.query.RecordsQueryResult;
import ru.citeck.ecos.records2.source.dao.local.LocalRecordsDao;
import ru.citeck.ecos.records2.source.dao.local.MutableRecordsLocalDao;
import ru.citeck.ecos.records2.source.dao.local.v2.LocalRecordsMetaDao;
import ru.citeck.ecos.records2.source.dao.local.v2.LocalRecordsQueryWithMetaDao;
import ru.citeck.ecos.records3.record.atts.schema.annotation.AttName;

import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.stream.Collectors;

@Component
public class NumTemplateRecordsDao extends LocalRecordsDao
    implements LocalRecordsMetaDao<NumTemplateRecordsDao.NumTemplateRecord>,
    LocalRecordsQueryWithMetaDao<NumTemplateRecordsDao.NumTemplateRecord>,
    MutableRecordsLocalDao<NumTemplateRecordsDao.NumTemplateRecord> {

    private static final String ID = "num-template";
    private final NumTemplateService numTemplateService;
    private final RecordEventsService recordEventsService;

    public NumTemplateRecordsDao(NumTemplateService numTemplateService, RecordEventsService recordEventsService) {
        setId(ID);
        this.numTemplateService = numTemplateService;
        this.recordEventsService = recordEventsService;

        numTemplateService.addListener(this::onTemplateChanged);
    }

    @Override
    public List<NumTemplateRecord> getValuesToMutate(List<RecordRef> records) {
        return getLocalRecordsMeta(records, null);
    }

    @Override
    public RecordsQueryResult<NumTemplateRecord> queryLocalRecords(RecordsQuery recordsQuery, MetaField metaField) {

        RecordsQueryResult<NumTemplateRecord> result = new RecordsQueryResult<>();

        if ("predicate".equals(recordsQuery.getLanguage())) {

            Predicate predicate = recordsQuery.getQuery(Predicate.class);

            Collection<EntityWithMeta<NumTemplateDef>> types = numTemplateService.getAll(
                recordsQuery.getMaxItems(),
                recordsQuery.getSkipCount(),
                predicate,
                LegacyRecordsUtils.mapLegacySortBy(recordsQuery.getSortBy())
            );

            result.setRecords(types.stream()
                .map(NumTemplateRecord::new)
                .collect(Collectors.toList()));
            result.setTotalCount(numTemplateService.getCount(predicate));
            return result;
        }

        if ("criteria".equals(recordsQuery.getLanguage())) {

            result.setRecords(numTemplateService.getAll(recordsQuery.getMaxItems(), recordsQuery.getSkipCount())
                .stream()
                .map(NumTemplateRecord::new)
                .collect(Collectors.toList())
            );
            result.setTotalCount(numTemplateService.getCount());

            return result;
        }

        return new RecordsQueryResult<>();
    }

    @Override
    public RecordsMutResult save(List<NumTemplateRecord> values) {

        RecordsMutResult result = new RecordsMutResult();
        values.forEach(dto -> {
            if (StringUtils.isBlank(dto.getId())) {
                throw new IllegalArgumentException("Attribute 'id' is mandatory");
            }
            String id = numTemplateService.save(dto).getEntity().getId();
            result.addRecord(new RecordMeta(id));
        });

        return result;
    }

    @Override
    public RecordsDelResult delete(RecordsDeletion deletion) {
        RecordsDelResult result = new RecordsDelResult();
        deletion.getRecords().stream()
            .map(RecordRef::getId)
            .forEach(id -> {
                numTemplateService.delete(id);
                result.addRecord(new RecordMeta(id));
            });
        return result;
    }

    @Override
    public List<NumTemplateRecord> getLocalRecordsMeta(List<RecordRef> records, MetaField metaField) {
        return records.stream()
            .map(RecordRef::getId)
            .map(id -> numTemplateService.getById(id).map(NumTemplateWithMetaDto::new)
                .orElseGet(() -> new NumTemplateWithMetaDto(id)))
            .map(NumTemplateRecord::new)
            .collect(Collectors.toList());
    }

    private void onTemplateChanged(@Nullable EntityWithMeta<NumTemplateDef> before,
                                   @Nullable EntityWithMeta<NumTemplateDef> after) {
        if (after != null) {
            recordEventsService.emitRecChanged(
                before,
                after,
                getId(),
                dto -> new NumTemplateRecord(new NumTemplateWithMetaDto(dto))
            );
        }
    }

    @NoArgsConstructor
    public static class NumTemplateRecord extends NumTemplateWithMetaDto {

        @Getter @Setter
        private List<String> modelAttributes;

        public NumTemplateRecord(NumTemplateWithMetaDto model) {
            super(model);
            modelAttributes = new ArrayList<>(TmplUtils.getAtts(this.getCounterKey()));
        }

        public NumTemplateRecord(EntityWithMeta<NumTemplateDef> model) {
            super(model);
            modelAttributes = model.getEntity().getModelAttributes();
        }

        @AttName("_type")
        public RecordRef getEcosType() {
            return RecordRef.create(EcosModelApp.NAME, "type", "number-template");
        }

        public String getModuleId() {
            return getId();
        }

        public void setModuleId(String value) {
            setId(value);
        }

        @AttName("?disp")
        public String getDisplayName() {
            return getId();
        }

        @JsonProperty("_content")
        public void setContent(List<ObjectData> content) {

            String dataUriContent = content.get(0).get("url", "");
            ObjectData data = Json.getMapper().read(dataUriContent, ObjectData.class);

            Json.getMapper().applyData(this, data);
        }

        @JsonValue
        @com.fasterxml.jackson.annotation.JsonValue
        public NumTemplateDto toJson() {
            return new NumTemplateDto(this);
        }

        public byte[] getData() {
            return YamlUtils.toNonDefaultString(toJson()).getBytes(StandardCharsets.UTF_8);
        }
    }
}

File: src/main/java/ru/citeck/ecos/model/num/api/records/NumTemplateRecordsDao.kt
package ru.citeck.ecos.model.num.api.records

import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonValue
import org.springframework.stereotype.Component
import ru.citeck.ecos.commons.data.ObjectData
import ru.citeck.ecos.commons.data.entity.EntityWithMeta
import ru.citeck.ecos.commons.json.Json.mapper
import ru.citeck.ecos.commons.json.YamlUtils.toNonDefaultString
import ru.citeck.ecos.commons.utils.TmplUtils.getAtts
import ru.citeck.ecos.context.lib.auth.AuthContext
import ru.citeck.ecos.events2.type.RecordEventsService
import ru.citeck.ecos.model.lib.num.dto.NumTemplateDef
import ru.citeck.ecos.model.lib.workspace.IdInWs
import ru.citeck.ecos.model.lib.workspace.WorkspaceService
import ru.citeck.ecos.model.num.dto.NumTemplateDto
import ru.citeck.ecos.model.num.dto.NumTemplateWithMetaDto
import ru.citeck.ecos.model.num.service.NumTemplateService
import ru.citeck.ecos.records2.RecordConstants
import ru.citeck.ecos.records2.predicate.model.AndPredicate
import ru.citeck.ecos.records2.predicate.model.Predicate
import ru.citeck.ecos.records2.predicate.model.Predicates
import ru.citeck.ecos.records3.record.atts.dto.LocalRecordAtts
import ru.citeck.ecos.records3.record.atts.schema.annotation.AttName
import ru.citeck.ecos.records3.record.atts.value.factory.bean.BeanTypeUtils
import ru.citeck.ecos.records3.record.dao.AbstractRecordsDao
import ru.citeck.ecos.records3.record.dao.atts.RecordAttsDao
import ru.citeck.ecos.records3.record.dao.delete.DelStatus
import ru.citeck.ecos.records3.record.dao.delete.RecordsDeleteDao
import ru.citeck.ecos.records3.record.dao.mutate.RecordMutateDao
import ru.citeck.ecos.records3.record.dao.query.RecordsQueryDao
import ru.citeck.ecos.records3.record.dao.query.dto.query.RecordsQuery
import ru.citeck.ecos.records3.record.dao.query.dto.res.RecsQueryRes
import ru.citeck.ecos.webapp.api.constants.AppName
import ru.citeck.ecos.webapp.api.entity.EntityRef
import java.nio.charset.StandardCharsets

@Component
class NumTemplateRecordsDao(
    private val numTemplateService: NumTemplateService,
    private val recordEventsService: RecordEventsService,
    private val workspaceService: WorkspaceService
) : AbstractRecordsDao(),
    RecordAttsDao,
    RecordsQueryDao,
    RecordsDeleteDao,
    RecordMutateDao {

    companion object {
        private const val ID = "num-template"
    }

    init {
        numTemplateService.addListener { before, after ->
            this.onTemplateChanged(before, after)
        }
    }

    override fun queryRecords(recsQuery: RecordsQuery): Any? {
        val result = RecsQueryRes<NumTemplateRecord>()

        if ("predicate" == recsQuery.language) {
            var predicate = recsQuery.getQuery(
                Predicate::class.java
            )
            if (recsQuery.workspaces.isNotEmpty()) {
                val newPredicate = AndPredicate()
                newPredicate.addPredicate(predicate)
                newPredicate.addPredicate(
                    Predicates.inVals(
                        "workspace",
                        recsQuery.workspaces.map {
                            if (workspaceService.isWorkspaceWithGlobalEntities(it)) "" else it
                        }
                    )
                )
                predicate = newPredicate
            }

            val types = numTemplateService.getAll(
                recsQuery.page.maxItems,
                recsQuery.page.skipCount,
                predicate,
                recsQuery.sortBy
            )

            result.setRecords(types.map { NumTemplateRecord(it) })
            result.setTotalCount(numTemplateService.getCount(predicate))
            return result
        }

        if ("criteria" == recsQuery.language) {
            result.setRecords(
                numTemplateService.getAll(
                    recsQuery.page.maxItems,
                    recsQuery.page.skipCount
                ).map { NumTemplateRecord(it) }
            )

            result.setTotalCount(numTemplateService.getCount())

            return result
        }

        return RecsQueryRes<Any>()
    }

    override fun mutate(record: LocalRecordAtts): String {
        val dto = getRecordAtts(record.id)

        val dtoCtx = BeanTypeUtils.getTypeContext(dto::class.java)
        dtoCtx.applyData(dto, record.attributes)

        require(dto.id.isNotBlank()) { "Attribute 'id' is mandatory" }

        checkMutPerms(dto)

        val resEntity = numTemplateService.save(dto).entity
        return workspaceService.addWsPrefixToId(resEntity.id, resEntity.workspace)
    }

    private fun checkMutPerms(record: NumTemplateRecord) {
        if (AuthContext.isRunAsSystemOrAdmin()) {
            return
        }
        if (record.workspace.isNotBlank()) {
            if (workspaceService.isUserManagerOf(AuthContext.getCurrentUser(), record.workspace)) {
                return
            } else {
                error("Permission denied. You can't create number templates in workspace '${record.workspace}'")
            }
        } else {
            error("Permission denied. You can't create number templates in global scope.")
        }
    }

    override fun delete(recordIds: List<String>): List<DelStatus> {
        val results: MutableList<DelStatus> = ArrayList()

        for (recordId in recordIds) {
            numTemplateService.delete(workspaceService.convertToIdInWs(recordId))
            results.add(DelStatus.OK)
        }
        return results
    }

    override fun getRecordAtts(recordId: String): NumTemplateRecord {
        val idInWs = workspaceService.convertToIdInWs(recordId)
        val dto = numTemplateService.getByIdOrNull(idInWs)?.let {
            NumTemplateWithMetaDto(it)
        } ?: NumTemplateWithMetaDto(recordId)

        return NumTemplateRecord(dto)
    }

    private fun onTemplateChanged(
        before: EntityWithMeta<NumTemplateDef>?,
        after: EntityWithMeta<NumTemplateDef>?
    ) {
        if (after != null) {
            recordEventsService.emitRecChanged(
                before,
                after,
                getId()
            ) { dto -> NumTemplateRecord(NumTemplateWithMetaDto(dto)) }
        }
    }

    override fun getId(): String {
        return ID
    }

    inner class NumTemplateRecord : NumTemplateWithMetaDto {

        private var modelAttributes: List<String>

        private val originalId = this.id

        constructor(model: NumTemplateWithMetaDto) : super(model) {
            modelAttributes = ArrayList(getAtts(this.counterKey))
        }

        constructor(model: EntityWithMeta<NumTemplateDef>) : super(model) {
            modelAttributes = model.entity.modelAttributes
        }

        fun getModelAttributes(): List<String> {
            return modelAttributes
        }

        @JsonProperty(RecordConstants.ATT_WORKSPACE)
        fun setCtxWorkspace(workspace: String?) {
            if (originalId != this.id) {
                this.workspace = workspace ?: ""
            } else {
                this.workspace = workspaceService.getUpdatedWsInMutation(this.workspace, workspace)
            }
        }

        @AttName("?id")
        fun getRef(): EntityRef {
            return EntityRef.create(
                AppName.EMODEL,
                ID,
                workspaceService.convertToStrId(IdInWs.create(workspace, id))
            )
        }

        @AttName("_type")
        fun getEcosType(): EntityRef {
            return EntityRef.create(AppName.EMODEL, "type", "number-template")
        }

        @AttName("?disp")
        fun getDispName(): String {
            return name.ifBlank { id }
        }

        @JsonProperty("_content")
        fun setContent(content: List<ObjectData>) {
            val dataUriContent: String = content.first().get("url", "")
            val data = mapper.read(dataUriContent, ObjectData::class.java)
            mapper.applyData(this, data)
        }

        @JsonValue
        fun toJson(): NumTemplateDto {
            val dto = NumTemplateDto(this)
            dto.workspace = ""
            return dto
        }

        fun getData(): ByteArray {
            return toNonDefaultString(toJson()).toByteArray(StandardCharsets.UTF_8)
        }
    }
}
